runtime/sys.c

```diff
@@ -62,9 +62,23 @@
 #include "caml/major_gc.h"
 #include "caml/shared_heap.h"
 
-static char * error_message(void)
+CAMLexport char * caml_strerror(int errnum, char * buf, size_t buflen)
 {
-  return strerror(errno);
+#ifdef _WIN32
+  /* Windows has a thread-safe strerror */
+  return strerror(errnum);
+#elif defined(_GNU_SOURCE)
+  return strerror_r(errnum, buf, buflen);
+#else
```

Comment:
I'm not sure this special case is necessary: `_GNU_SOURCE` is not defined in this file, and the fallback case below would result in a type error if the GNU version of `strerror_r`  as , and if it were the C compiler would error on the use of `strerror_r` with the wrong type.

But I don't mind keeping the special case, and I checked that the function compiles and runs correctly whether `_GNU_SOURCE` is defined or not.

Reply:
Don't you want to define _GNU_SOURCE in order to get the GNU implementation, though? That would avoid a data copy for each caml_strerror call, since GNU strerror_r usually just returns an immutable pointer to its static array of error messages (whereas the XSI version is forced to copy into the destination buffer).  That's why I stuck it in a separate file in the other PR, for header hygiene.

Reply:
Ah, now I understand the motivation for the special case: it's not "do the right thing if _GNU_SOURCE happens to be defined", it's "use the GNU version for performance".  Sorry for having missed that.  I need to think about it some more, but globally I'm not too concerned about the extra copy done by the POSIX `strerror_r`.

Reply:
I went out on a limb and removed support for the GNU version of `strerror_r`.  It feels safer to rely exclusively on the POSIX function.  Nothing is performance-critical here.  Let me know what you think.